src/data/scripts/campaign/colonies/VayraColonialManager.java

```diff
@@ -254,88 +304,104 @@ public void advance(float amount) {
-                    return;
-                }
-                spamLog("VayraColonialManager::advance()\tUPGRADE SECTION\ttarget.isPlanetConditionMarketingOnly ? "+target.isPlanetConditionMarketOnly());
-                if (!target.isPlanetConditionMarketOnly()) {
-                    log.info(String.format("We were gonna colonize %s, but it's not a planetary condition only market "
-                            + "(already taken?) so we're giving up instead", target.getName()));
-                    spamLog("VayraColonialManager::advance()\tUPGRADE SECTION\taborting because !target.isPlanetConditionMarketOnly() condition");
-                }
-                float fleetPoints = pickExpeditionFP(colonyFaction);
-                log.info(String.format("Assembling %s colonial expedition at %s, target: %s", colonyFaction.getDisplayNameLong(), source.getName(), target.getName()));
-                try {
-                    VayraColonialExpeditionIntel expedition = new VayraColonialExpeditionIntel(colonyFaction, source, target, fleetPoints);
-                    planetsTargetedForColonies.add(target);
-                    spamLog("VayraColonialManager::advance()\tUPGRADE SECTION\tadded planet as target for colony\ttarget name: "+target.getName()+", target star system: "+target.getStarSystem());
-                } catch (NullPointerException ex) {
-                    log.info(String.format("Expedition picked %s in %s as a target, but that causes an error -- i thought i fixed this...", target.getName(), target.getStarSystem()));
-                }
-            }
+            performColonySpawn(colonyChance);
         }
     }
 
-    public String stringifyColonyList() {
-        StringBuffer sb = new StringBuffer();
-        for (Iterator<MarketAPI> iter = coloniesActive.iterator(); iter.hasNext();  ) {
-            MarketAPI colony = iter.next();
-            sb
-                    .append("Faction ID: ")
-                    .append(colony.getFactionId())
-                    .append(", ");
+    private void performColonyUpgrade() {
+        spamLog("VayraColonialManager::performColonyUpgrade()\tUPGRADE SECTION\tcoloniesActive size: "+coloniesActive.size());
+        if (!coloniesActive.isEmpty()) {
+            for (MarketAPI market : coloniesActive) {
+                pickNextUpgrade(market);
+            }
+            List<MarketAPI> remove = new ArrayList<>();
+            for (MarketAPI toRemove : coloniesToRemove) {
+                coloniesActive.remove(toRemove);
+                log.info(String.format("removing %s from list of colonies as it now belongs to %s", toRemove.getName(), toRemove.getFactionId()));
+                purgeCores(toRemove);
+                executeAdmin(toRemove);
+                remove.add(toRemove);
+            }
+            for (MarketAPI toRemove : remove) {
+                coloniesToRemove.remove(toRemove);
+            }
+
+            if (AI_REBELLION_THRESHOLD > 0) {
```

Comment:
something tells me this should be exposed as well in the "colonial factions" section